middleman_protocol/middleman_protocol/message.py

```diff
@@ -0,0 +1,387 @@
+from golem_messages.cryptography import ecdsa_verify
+from golem_messages.exceptions import InvalidSignature
+from golem_messages.message.base import Message as BaseGolemMessage
+
+from .constants import ESCAPE_CHARACTER
+from .constants import ESCAPE_SEQUENCES
+from .constants import FRAME_PAYLOAD_LENGTH_BYTES_LENGTH
+from .constants import FRAME_PAYLOAD_LENGTH_STARTING_BYTE
+from .constants import FRAME_PAYLOAD_STARTING_BYTE
+from .constants import FRAME_REQUEST_ID_BYTES_LENGTH
+from .constants import FRAME_SEPARATOR
+from .constants import FRAME_SEPARATOR_LENGTH
+from .constants import FRAME_SIGNATURE_BYTES_LENGTH
+from .constants import PayloadType
+from .exceptions import FrameInvalidMiddlemanProtocolError
+from .exceptions import PayloadInvalidMiddlemanProtocolError
+from .exceptions import PayloadLengthMismatchMiddlemanProtocolError
+from .exceptions import PayloadTypeInvalidMiddlemanProtocolError
+from .exceptions import RequestIdInvalidTypeMiddlemanProtocolError
+from .exceptions import SignatureInvalidMiddlemanProtocolError
+from .registry import PAYLOAD_TYPE_TO_MIDDLEMAN_MESSAGE_CLASS
+from .registry import register
+
+
+class AbstractMiddlemanMessage(ABC):
+    """
+    Abstract class for inheriting by Middleman protocol messages.
+
+    Contains base functionality like:
+    - serialization,
+    - deserialization,
+    - frame validation.
+
+    Subclasses must implement:
+    - parsing payload to bytes on serialization,
+    - parsing serialized bytes payload back to original format on deserialization,
+    - validation of original payload before serialization.
+
+    """
+
+    __slots__ = [
+        'payload',
+        'payload_type',
+        'request_id',
+    ]
+
+    def __init__(self, payload: Any, request_id: int) -> None:
+        self._validate_request_id(request_id)
+        self.request_id = request_id
+
+        self._validate_payload(payload)
+        self.payload = payload
+
+    @classmethod
+    @abstractmethod
+    def _deserialize_payload(cls, payload: bytes):
+        pass
+
+    @abstractmethod
+    def _serialize_payload(self, payload: Any):
+        pass
+
+    @abstractmethod
+    def _validate_payload(self, payload: Any):
+        pass
+
+    @classmethod
+    def get_frame_format(cls) -> Struct:
+        """ Returns Struct object containing frame structure. """
+        frame_format = Struct(
+            frame_signature=Bytes(FRAME_SIGNATURE_BYTES_LENGTH),
+            signed_part_of_the_frame=Struct(
+                request_id=BytesInteger(FRAME_REQUEST_ID_BYTES_LENGTH),
+                payload_type=Enum(Byte, PayloadType),
+                payload_length=BytesInteger(FRAME_PAYLOAD_LENGTH_BYTES_LENGTH),
+                payload=GreedyBytes,
+            ),
+        )
+
+        assert FRAME_PAYLOAD_LENGTH_STARTING_BYTE == (
+            frame_format.frame_signature.subcon.length +
+            frame_format.signed_part_of_the_frame.request_id.subcon.length +
+            1  # 1 byte for payload_type
```

Comment:
Maybe `PAYLOAD_TYPE_LENGTH` would be a good idea? It's just 1 so not a big issue, but still, this should probably have a constant.

Reply:
:+1: done, but it can't be used in enum contruct in frame. At first (I have not examined it more than what is in documentation examples) it looks like there can't be enum with more than byte.

Reply:
In that case maybe `CONSTRUCT_ENUM_LENGTH`? Then you would not have to create a new constant for every enum.

Anyway, it's not a big deal. 1 would be fine too, though a constant at least explains why it's 1.

Reply:
But you can't use it enum definition. So far what I have understood you can only use `Byte` type from construct. It failed for me even with `Bytes(1)` (which I wanted to replace with constant).

So after all I have left it to be `PAYLOAD_TYPE_LENGTH`.